### ISSUE TITLE
ocaml-dune-configurator 3.4.1 fails to build on MacOS: Library "csexp" not found

### ISSUE THREAD

**Issue by barracuda156** [ocaml/dune#6156]
`ocaml-dune-configurator` 3.4.1 fails to build on MacOS, since it fails to find `ocaml-csexp` library:
{{{
--->  Configuring ocaml-dune-configurator
--->  Building ocaml-dune-configurator
Executing:  cd "/opt/local/var/macports/build/_opt_PPCRosettaPorts_ocaml_ocaml-dune/ocaml-dune-configurator/work/dune-3.4.1" && /opt/local/bin/dune build -j 6 --root=/opt/local/var/macports/build/_opt_PPCRosettaPorts_ocaml_ocaml-dune/ocaml-dune-configurator/work/dune-3.4.1 --profile=release --ignore-promoted-rules --no-config --only-packages=dune-configurator --default-target=@install @install 
File "otherlibs/configurator/src/dune", line 7, characters 17-22:
7 |  (libraries unix csexp)
                     ^^^^^
Error: Library "csexp" not found.
-> required by library "dune-configurator" in
   _build/default/otherlibs/configurator/src
-> required by _build/default/META.dune-configurator
-> required by _build/install/default/lib/dune-configurator/META
-> required by _build/default/dune-configurator.install
-> required by alias install
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCRosettaPorts_ocaml_ocaml-dune/ocaml-dune-configurator/work/dune-3.4.1" && /opt/local/bin/dune build -j 6 --root=/opt/local/var/macports/build/_opt_PPCRosettaPorts_ocaml_ocaml-dune/ocaml-dune-configurator/work/dune-3.4.1 --profile=release --ignore-promoted-rules --no-config --only-packages=dune-configurator --default-target=@install @install 
Exit code: 1
}}}

The same failure occurs on newest versions of MacOS as well as on old ones. Logs from MacOS 11 here: https://github.com/macports/macports-ports/pull/14919

**Comment by emillon** [comment]
Hi! configurator depends on csexp. This is recorded [in the upstream opam file](https://github.com/ocaml/dune/blob/5df2221179ff67e51ab93334ab2f8e42cdffb95b/dune-configurator.opam#L24) as well as [in opam-repository](https://github.com/ocaml/opam-repository/blob/3fd9653f4de083bc23adc5306f3c7fb18239d7fe/packages/dune-configurator/dune-configurator.3.4.1/opam#L23). This seems to be mentioned in the port file, and the logs mention that it is being installed:

> Installing dependency (1 of 1) 'ocaml-csexp' with variants '' ... [OK]

So that looks like an issue on the macports side. Something I'd check is whether csexp has installed any files. Another thing to note is that dune 3.4.1 embeds a copy of csexp in its tarball, which is deleted before building configurator (see previously linked files). But I don't think that's the issue.

**Comment by emillon** [comment]
Hum, actually those are arguments to `dune install`. Still might be relevant depending on what you use to install packages in macports.

**Comment by barracuda156** [comment]
> Hum, actually those are arguments to `dune install`. Still might be relevant depending on what you use to install packages in macports.

`ocaml-csexp` is a dependency and it is installed, of course:
```
subport ocaml-dune-configurator {
    ocaml.build_type    dune
    depends_lib-append  port:ocaml-csexp
}
```
Its portfile: https://github.com/macports/macports-ports/blob/master/ocaml/ocaml-csexp/Portfile

`ocaml.build_type` comes from `ocaml` PG: https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/ocaml-1.1.tcl

Updated portfile for `ocaml-dune`: https://github.com/barracuda156/macports-ports/blob/ocaml-dune/ocaml/ocaml-dune/Portfile
(`ocaml-dune` itself builds and works, it is `ocaml-dune-configurator` that fails.)

**Comment by emillon** [comment]
Let me rephrase what I mean - the Macports PR changes configure arguments. I suspect that this causes `dune` not to be able to find installed packages at all (which would not be visible when building csexp, because it has no ocaml lib dependencies). What happens when you don't change the flags?

**Comment by barracuda156** [comment]
> Let me rephrase what I mean - the Macports PR changes configure arguments. I suspect that this causes `dune` not to be able to find installed packages at all (which would not be visible when building csexp, because it has no ocaml lib dependencies). What happens when you don't change the flags?

@emillon I will try that, thank you for suggestion. (I did not blame PortGroup initially, since it does work fine with many other OCaml ports, I have verified that locally.)

**Comment by barracuda156** [comment]
@emillon It works a bit differently. All ports using `ocaml` PG get default dependencies automatically, unless something is set to be overridden in portfile. I can see that `ocaml-csexp` installs normally, has all files, has dune-package, has public_name. It is visible to the environment:
```
36-84% ocamlfind query csexp
/opt/local/lib/ocaml/site-lib/csexp
36-84% dune installed-libraries
bigarray                     (version: [distributed with OCaml])
bigarray-compat              (version: 1.0.0)
biniou                       (version: n/a)
bytes                        (version: [distributed with OCaml])
camlp4                       (version: 4.14.0)
camlp4.exceptiontracer       (version: 4.14.0)
camlp4.extend                (version: 4.14.0)
camlp4.foldgenerator         (version: 4.14.0)
camlp4.fulllib               (version: 4.14.0)
camlp4.gramlib               (version: 4.14.0)
camlp4.lib                   (version: 4.14.0)
camlp4.listcomprehension     (version: 4.14.0)
camlp4.locationstripper      (version: 4.14.0)
camlp4.macro                 (version: 4.14.0)
camlp4.mapgenerator          (version: 4.14.0)
camlp4.metagenerator         (version: 4.14.0)
camlp4.profiler              (version: 4.14.0)
camlp4.quotations            (version: 4.14.0)
camlp4.quotations.o          (version: 4.14.0)
camlp4.quotations.r          (version: 4.14.0)
camlp4.tracer                (version: 4.14.0)
cmdliner                     (version: n/a)
compiler-libs                (version: [distributed with OCaml])
compiler-libs.bytecomp       (version: [distributed with OCaml])
compiler-libs.common         (version: [distributed with OCaml])
compiler-libs.toplevel       (version: [distributed with OCaml])
csexp                        (version: n/a)
dune.configurator            (version: n/a)
dynlink                      (version: [distributed with OCaml])
easy-format                  (version: n/a)
fileutils                    (version: n/a)
fileutils.str                (version: n/a)
findlib                      (version: 1.9.3)
findlib.dynload              (version: 1.9.3)
findlib.internal             (version: 1.9.3)
findlib.top                  (version: 1.9.3)
fmt                          (version: %%VERSION_NUM%%)
fmt.cli                      (version: %%VERSION_NUM%%)
fmt.top                      (version: %%VERSION_NUM%%)
fmt.tty                      (version: %%VERSION_NUM%%)
jane-street-headers          (version: v0.15.0)
menhirLib                    (version: 20210310)
menhirSdk                    (version: 20210310)
oUnit                        (version: 2.2.4)
oUnit.advanced               (version: n/a)
oUnit.threads                (version: n/a)
ocaml-compiler-libs.bytecomp (version: n/a)
ocaml-compiler-libs.common   (version: n/a)
ocaml-compiler-libs.shadow   (version: n/a)
ocaml-compiler-libs.toplevel (version: n/a)
ocaml-migrate-parsetree      (version: 2.3.0)
ocamlbuild                   (version: 0.14.1)
ocamldoc                     (version: [distributed with OCaml])
ounit2                       (version: n/a)
ounit2.advanced              (version: n/a)
ounit2.threads               (version: n/a)
ppx_derivers                 (version: n/a)
ppx_deriving                 (version: n/a)
ppx_deriving.api             (version: 5.2.1)
ppx_deriving.create          (version: 5.2.1)
ppx_deriving.enum            (version: 5.2.1)
ppx_deriving.eq              (version: 5.2.1)
ppx_deriving.fold            (version: 5.2.1)
ppx_deriving.iter            (version: 5.2.1)
ppx_deriving.make            (version: 5.2.1)
ppx_deriving.map             (version: 5.2.1)
ppx_deriving.ord             (version: 5.2.1)
ppx_deriving.runtime         (version: 5.2.1)
ppx_deriving.show            (version: 5.2.1)
ppx_deriving.std             (version: 5.2.1)
ppx_deriving_yojson          (version: 3.7.0)
ppx_deriving_yojson.runtime  (version: 3.7.0)
ppx_ignore_instrumentation   (version: v0.15.0)
ppx_pipebang                 (version: v0.15.0)
ppx_tools                    (version: 6.6)
ppx_tools.ast_lifter         (version: 6.6)
ppx_tools.metaquot           (version: 6.6)
ppxlib                       (version: n/a)
ppxlib.ast                   (version: n/a)
ppxlib.astlib                (version: n/a)
ppxlib.metaquot              (version: n/a)
ppxlib.metaquot_lifters      (version: n/a)
ppxlib.print_diff            (version: n/a)
ppxlib.runner                (version: n/a)
ppxlib.runner_as_ppx         (version: n/a)
ppxlib.stdppx                (version: n/a)
ppxlib.traverse              (version: n/a)
ppxlib.traverse_builtins     (version: n/a)
re                           (version: n/a)
re.emacs                     (version: n/a)
re.glob                      (version: n/a)
re.pcre                      (version: n/a)
re.perl                      (version: n/a)
re.posix                     (version: n/a)
re.str                       (version: n/a)
result                       (version: n/a)
seq                          (version: 0.3.1)
sexplib0                     (version: v0.14.0)
stdlib                       (version: [distributed with OCaml])
stdlib-shims                 (version: n/a)
str                          (version: [distributed with OCaml])
threads                      (version: [distributed with OCaml])
threads.posix                (version: [distributed with OCaml])
topkg                        (version: 1.0.3)
unix                         (version: [distributed with OCaml])
yojson                       (version: n/a)
yojson.biniou                (version: n/a)
```
What is broken is `dune-configurator`, which somehow still fails to see the lib. As you can see, quite a few packages are installed, so it is not that Macports is configured in a wrong way. Something either is wrong on `ocaml-csexp` side or Dune.

@rgrinberg Guys help me please, this silly issue is holding back a big project and I wasted several hours looking for solution. Similar cases were reported occasionally, but no fix.

It probably should just work, like with most of other packages using OCaml. If there are some complexities with Dune specifically or if `ocaml-csexp` is defective, maybe the functionality can be integrated?

**Comment by emillon** [comment]
I bisected and I confirm that it's the cause.
The scenario I bisected is:
- checkout a dune version
- install it
- build and install csexp using it
- `dune build -p dune-configurator`

This works on 2.0.1, fails on 3.6.0. This started to fail between 5aa4adacc3252f58d6dfafd183 and 8c3a2eddf8781d70cc96b6b808b45bbdfb5b833e. But that sort of makes sense: at that point, the build instructions in the opam files [started to mention removing the vendored copies](https://github.com/ocaml/dune/blame/361c46d83b3daa8fd3959afa4c83f01d0422b349/dune-configurator.opam#L30-L31).

If I read the portfile and ocaml-1.1.tcl correctly, you're setting a build type of "dune", which ignores the build instructions in the opam file and hardcodes dune instructions - that's not correct any more to build configurator.3.6.0.

**Comment by barracuda156** [comment]
@emillon What would be a correct solution?

I have been able to build it locally be using vendored copy instead of external lib dependency. However it still failed on Macports’ buildbot, which I am yet to figure out the reason for.

Sources still define `csexp` as a local lib, which is arguably wrong, if local one is destroyed by build script.

**Comment by emillon** [comment]
The right way is to remove the vendored directories after unpacking the tarball (in `post-extract` if I understand correctly). `csexp` is packaged separately and has its own tarball - the version embedded in the dune tarball shouldn't be installed. For example, end users should be able to upgrade `csexp` independently from `dune`.

**Comment by barracuda156** [comment]
> The right way is to remove the vendored directories after unpacking the tarball (in `post-extract` if I understand correctly). `csexp` is packaged separately and has its own tarball - the version embedded in the dune tarball shouldn't be installed. For example, end users should be able to upgrade `csexp` independently from `dune`.

This is what the code has been doing, and it exactly does not work.

**Comment by rgrinberg** [comment]
Does your configurator build script delete the vendor directory?

**Comment by barracuda156** [comment]
> Does your configurator build script delete the vendor directory?

I believe this is what it is set to do: https://github.com/ocaml/dune/blob/557a001436beaa3e4424e79b52a81c23b4e75710/dune-configurator.opam#L28

And what I started with was existing Macports portfile from Dune 2.x, which was absolutely simple: https://github.com/macports/macports-ports/blob/3c1ace9d14ed87685e8104301ff091df3519a97d/ocaml/ocaml-dune/Portfile#L33 (it does not modify sources and just uses Dune to build dune-configurator). That worked with Dune 2.x and stopped working with Dune 3.x: build system cannot find `csexp`, even though it is installed and visible: https://github.com/ocaml/dune/issues/6156#issuecomment-1320610206

**Comment by rgrinberg** [comment]
Given this "portfile" for configurator https://github.com/macports/macports-ports/blob/3c1ace9d14ed87685e8104301ff091df3519a97d/ocaml/ocaml-configurator/Portfile#L23

Could you point me to the line that would be responsible for deleting the vendor directory?

**Comment by barracuda156** [comment]
@rgrinberg Portfile does not set anything to be deleted from the sources. By setting `ocaml.build_type dune` it passes handling of the build to Dune. No custom settings are added for the subport. I believe, that Dune is then deleting vendored dirs, since that is what I see in its build files in the source. (OCaml is not an area of my expertise, so please forgive me if I miss something obvious.)

If you can recommend any specific way to fix the build (given Macports environment, but it is pretty flexible), I will try that. What worked before with Dune 2 does not work now.

Setting available via `ocaml` port group are listed here: https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/ocaml-1.1.tcl

**Comment by barracuda156** [comment]
@rgrinberg Okay, I will try that and update in a while.

**Comment by barracuda156** [comment]
@rgrinberg Sorry for a forgotten topic, it worked, thank you.
